test/pipelines/unit/test_pipeline.py

```diff
@@ -24,3 +24,300 @@ def test_max_loops():
+        "max_loops_allowed": 100,
+        "components": {
+            "add_two": {
+                "type": "AddFixedValue",
+                "hash": "123",
+                "init_parameters": {"add": 2},
+            },
+            "add_default": {
+                "type": "AddFixedValue",
+                "hash": "456",
+                "init_parameters": {},
+            },
+            "double": {
+                "type": "Double",
+                "hash": "789",
+                "init_parameters": {},
+            },
+        },
+        "connections": [
+            {"sender": "add_two.result", "receiver": "double.value"},
+            {"sender": "double.value", "receiver": "add_default.value"},
+        ],
+    }
+    pipe = Pipeline.from_dict(data)
+
+    assert pipe.metadata == {"test": "test"}
+    assert pipe.max_loops_allowed == 100
+
+    # Components
+    assert len(pipe.graph.nodes) == 3
+    ## add_two
+    add_two = pipe.graph.nodes["add_two"]
+    assert add_two["instance"].add == 2
+    assert add_two["input_sockets"] == {
+        "value": InputSocket(name="value", type=int, is_optional=False, sender=None),
+        "add": InputSocket(name="add", type=Optional[int], is_optional=True, sender=None),
+    }
+    assert add_two["output_sockets"] == {
+        "result": OutputSocket(name="result", type=int),
+    }
+    assert add_two["visits"] == 0
+
+    ## add_default
+    add_default = pipe.graph.nodes["add_default"]
+    assert add_default["instance"].add == 1
+    assert add_default["input_sockets"] == {
+        "value": InputSocket(name="value", type=int, is_optional=False, sender="double"),
+        "add": InputSocket(name="add", type=Optional[int], is_optional=True, sender=None),
+    }
+    assert add_default["output_sockets"] == {
+        "result": OutputSocket(name="result", type=int),
+    }
+    assert add_default["visits"] == 0
+
+    ## double
+    double = pipe.graph.nodes["double"]
+    assert double["instance"]
+    assert double["input_sockets"] == {
+        "value": InputSocket(name="value", type=int, is_optional=False, sender="add_two"),
+    }
+    assert double["output_sockets"] == {
+        "value": OutputSocket(name="value", type=int),
+    }
+    assert double["visits"] == 0
+
+    # Connections
+    connections = list(pipe.graph.edges(data=True))
+    assert len(connections) == 2
+    assert connections[0] == (
+        "add_two",
+        "double",
+        {
+            "conn_type": "int",
+            "from_socket": OutputSocket(name="result", type=int),
+            "to_socket": InputSocket(name="value", type=int, is_optional=False, sender="add_two"),
+        },
+    )
+    assert connections[1] == (
+        "double",
+        "add_default",
+        {
+            "conn_type": "int",
+            "from_socket": OutputSocket(name="value", type=int),
+            "to_socket": InputSocket(name="value", type=int, is_optional=False, sender="double"),
+        },
+    )
```

Comment:
Couldn't we rather build an identical pipeline and then do pipeline equality? We overrode `Pipeline.__eq__` precisely for this usecase I believe. If it doesn't work I'd rather fix `__eq__`

```suggestion
    add_two = AddFixedValue(add=2)
    add_default = AddFixedValue()
    double = Double()
    expected_pipe = Pipeline(metadata={"test": "test"})
    expected_pipe.add_component("add_two", add_two)
    expected_pipe.add_component("add_default", add_default)
    expected_pipe.add_component("double", double)
    expected_pipe.connect("add_two", "double")
    expected_pipe.connect("double", "add_default")
    
    assert pipe == expected_pipe
```

Reply:
I wouldn't use a simple equality comparison now to be fair, it's not as detailed as I want.
`Pipeline._comparable_nodes_list()` doesn't compare the component instances but only their classes. We miss some checks by simplifying like this.

I think it's best if we change this after we rework `__eq__` to make a more in-depth comparison. 
Though I would still be wary as changing `__eq__` could mean losing important checks in certain tests.

Same for the comment below.

Reply:
Alright! I think that if we make `__eq__` more precise other tests may only benefit at the moment. Let's do it later then